cpp/include/raft/sparse/solver/detail/lanczos.cuh

```diff
@@ -1396,4 +1438,658 @@ int computeLargestEigenvectors(
+                                       cusparse_u,
+                                       CUSPARSE_SPMV_ALG_DEFAULT,
+                                       cusparse_spmv_buffer.data_handle(),
+                                       stream);
+
+    auto alpha_i  = raft::make_device_scalar_view(&alpha(0, i));
+    auto v_vector = raft::make_device_vector_view<const value_type_t>(v.data_handle(), n);
+    auto u_vector = raft::make_device_vector_view<const value_type_t>(u.data_handle(), n);
+    raft::linalg::dot(handle, v_vector, u_vector, alpha_i);
+
+    raft::matrix::fill(handle, vv, zero);
+
+    auto cublas_h = resource::get_cublas_handle(handle);
+
+    value_type_t alpha_i_host = 0;
+    value_type_t b            = 0;
+    value_type_t mone         = -1;
+
+    raft::copy<value_type_t>(&b, &beta(0, (i - 1 + ncv) % ncv), 1, stream);
+    raft::copy<value_type_t>(&alpha_i_host, &(alpha(0, i)), 1, stream);
+
+    raft::linalg::detail::cublasaxpy(
+      cublas_h, n, &alpha_i_host, v.data_handle(), 1, vv.data_handle(), 1, stream);
+
+    raft::linalg::detail::cublasaxpy(
+      cublas_h, n, &b, &V((i - 1 + ncv) % ncv, 0), 1, vv.data_handle(), 1, stream);
+
+    raft::linalg::detail::cublasaxpy(
+      cublas_h, n, &mone, vv.data_handle(), 1, u.data_handle(), 1, stream);
+
+    raft::linalg::detail::cublasgemv(cublas_h,
+                                     CUBLAS_OP_T,
+                                     n,
+                                     i + 1,
+                                     &one,
+                                     V.data_handle(),
+                                     n,
+                                     u.data_handle(),
+                                     1,
+                                     &zero,
+                                     uu.data_handle(),
+                                     1,
+                                     stream);
+
+    raft::linalg::detail::cublasgemv(cublas_h,
+                                     CUBLAS_OP_N,
+                                     n,
+                                     i + 1,
+                                     &mone,
+                                     V.data_handle(),
+                                     n,
+                                     uu.data_handle(),
+                                     1,
+                                     &one,
+                                     u.data_handle(),
+                                     1,
+                                     stream);
+
+    auto uu_i = raft::make_device_scalar_view(&uu(0, i));
+    raft::linalg::add(handle, make_const_mdspan(alpha_i), make_const_mdspan(uu_i), alpha_i);
+
+    kernel_clamp_down<<<1, 1>>>(alpha_i.data_handle(), static_cast<value_type_t>(1e-9));
+
+    raft::linalg::detail::cublassetpointermode(cublas_h, CUBLAS_POINTER_MODE_DEVICE, stream);
+    raft::linalg::detail::cublasnrm2(cublas_h, n, u.data_handle(), 1, &beta(0, i), stream);
```

Comment:
Please expose any of these functions publicly if needed, but do not call into detail APIs.